### PR TITLE
Add support for differentiable weight decay

### PR DESCRIPTION
**Pull request by emmett-bicker**
(Actual) second PR in a larger project to broaden support for differentiable optimizers with @janeyx99!

In this PR, I did a lot of pattern matching from the previous PR to add support for differentiable weight_decay.

And also added a single new line on line 359 (previously line 352) to make the code from the last PR a little easier to read

Continuation of progress on #141832

### CODE REVIEW

**Comment by EmmettBicker** [comment]
Do you think that after we add differentiable support for another optimizer, it's time to look at step 0 and find a better testing framework? I'm feeling like these current tests will get very repetitive very quickly.

File: torch/optim/sgd.py [hunk 1]
```diff
@@ -334,7 +334,14 @@ def _single_tensor_sgd(
         grad = grads[i] if not maximize else -grads[i]
 
         if weight_decay != 0:
-            grad = grad.add(param, alpha=weight_decay)
+            # Nested if is necessary to bypass jitscript rules
+            if isinstance(weight_decay, Tensor):
+                if weight_decay.requires_grad:
```
Comment (janeyx99) [inline torch/optim/sgd.py:339]: can you add a comment here saying

# used for the differentiable path, which is why the param.clone() is needed

**Review by janeyx99 — APPROVED** [review]
Nice!

File: torch/optim/sgd.py [hunk 2]
```diff
@@ -334,7 +334,14 @@ def _single_tensor_sgd(
...
+                    grad = grad.addcmul_(param.clone(), weight_decay)
```
Comment (janeyx99) [inline torch/optim/sgd.py:340]:
```suggestion
                    # usually this is the differentiable path, which is why the param.clone() is needed
                    grad = grad.addcmul_(param.clone(), weight_decay)
```

**Comment by janeyx99** [comment]
> Do you think that after we add differentiable support for another optimizer, it's time to look at step 0 and find a better testing framework? I'm feeling like these current tests will get very repetitive very quickly.

Precisely :)

**Comment by EmmettBicker** [comment]
Ok, sounds great!!

**Comment by EmmettBicker** [comment]
Totally not urgent at all, but I'd love to know if you think this is ready for CI!

**Review by janeyx99 — APPROVED** [review]
